### PR TITLE
Support recurring events

### PR DESCRIPTION
**Pull request by fanwgwg**
Going to add more testcases

### CODE REVIEW

**Comment by burnflare** [comment]
LOL

**Comment by burnflare** [comment]
@fanwgwg Do fix the conflicts with master

**Comment by burnflare** [comment]
@fanwgwg Are you ready to merge this?

**Comment by fanwgwg** [comment]
@burnflare yup can merge, I'll just resolve the conflicts first

**Comment by INCENDE** [comment]
Btw weiguang are you applying the abstraction occurrence pattern?

**Comment by fanwgwg** [comment]
@INCENDE  yup I followed the pattern!

File: src/main/java/seedu/agendum/model/task/ChildRecurringTask.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
+package seedu.agendum.model.task;
+
+import java.time.LocalDateTime;
+import java.util.Optional;
+
+public class ChildRecurringTask extends RecurringTask {
```
Comment (INCENDE) [inline src/main/java/seedu/agendum/model/task/ChildRecurringTask.java:7]: I believe in the abstraction occurrence pattern, the occurence should not extend the abstraction

Comment (INCENDE) [reply src/main/java/seedu/agendum/model/task/ChildRecurringTask.java:7]: "The _Abstraction_ class should hold all common information, and the unique information
should be kept by the _Occurrence_ class" 
In damith's handout

Comment (fanwgwg) [reply src/main/java/seedu/agendum/model/task/ChildRecurringTask.java:7]: Yup it does say so on the handouts, but it never says the relationship between these two classes. In my implementation, RecurringTask stores the general information, childRecurringTask stores the unque information, I think it still follows the abstraction-occurrence pattern.

**Review by rachx — CHANGES_REQUESTED** [review]
Putting this on hold as I haven't read it thoroughly yet.
Can you explain the high level design e.g. how the model class diagram will change etc? (If not on github, then offline)

**Comment by fanwgwg** [comment]
@rachx  just updated the model class diagram

**Comment by rachx** [comment]
1) Should the creation of new occurrence be tied to mark?
1.1) Then what about unmarking? They should able to unmark as well.
2) How to delete an occurrence vs how to delete a recurrence? Is there a limit to the recurrence e.g. within a year?
3) can purposely add tasks (recurring or not) with the same properties as an existing recurring one.
4) now if i try to mark a recurring task occurrence multiple times, the new one will appear many times
5) how about if google calendar integration is successful, pull from there instead

**Comment by fanwgwg** [comment]
1.Of course this would be easier for users, otherwise how can the next task appear if the previous task hasn't bee completed yet?

1.1.These tasks are able to mark. But my idea is that, for example, if I have a recurring task to do every month, if I've already marked the Aug one, the Sep one, I can only unmark the latest one, which is the Sep one, instead of the Aug one, because it doesn't make sense to unmark the previous recurring task, and I will prompt to the user that "future tasks have been marked". If the latest occurence is unmarked, then all the previous occurence will go back one period, as well as the recurrence.

2.In my desigan, all occurence are completed, but the recurrence is never completed, because once the recurrence is marked, I will create new occurence of the same due time as the recurrence, and update the time of the recurrence to next recurring time. So delete a occurence, then it's just deleted, and it doesn't affect anything. If delete a recurrence, the recurring event is deleted, besides those occurrences that already marked.

3.This one I'll look into it.

4.Of course it should appear many times, because once you mark, its time will change, and all the appearred occurence are of different time. For example, you `add homework every day`, if you mark it once, time of this task will be changed to 31,Oct, and there will be a new task appearing in the completed task panel, which is 30,Oct. If you mark one more time, the recurrence will have time 1,Nov, and there will be one more occurence in the completed task panel, with time being 31, Oct.

5.It should be okay, cuz I just need to get the name and period of this recurring task, then just add into agendum.

Hope that makes sense.

**Comment by burnflare** [comment]
@rachx Are we dropping this?

**Comment by rachx** [comment]
Yes. Weiguang agreed

---- Vishnu Prem wrote ----

@rachxhttps://github.com/rachx Are we dropping this?

##

**Pull request closed without merging**